### ISSUE TITLE
tf.nn.tanh loses a finite first-derivative signal for float64 input

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#126637]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13

### Current behavior?

`tf.nn.tanh` returns zero for a first-derivative component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

For `tanh`, the derivative can be written as `4 * exp(-2 * abs(x)) / (1 + exp(-2 * abs(x)))**2`. At `x = -19.061547465398498`, the high-precision reference rounded to float64 is approximately `1.1102230246251531e-16`.

However, TensorFlow autodiff returns `0.0` for this derivative. The forward value rounds to `-1.0`, but the derivative remains representable in float64 and should not be lost during the first-order autodiff pass.

Expected behavior?

The first derivative of `tf.nn.tanh` at `x = -19.061547465398498` should be close to `1.1102230246251531e-16`, not `0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

x_value = -19.061547465398498
expected = 1.1102230246251531e-16
rtol = 1e-6

x = tf.Variable(x_value, dtype=tf.float64)

with tf.GradientTape() as tape:
    y = tf.nn.tanh(x)

g = tape.gradient(y, x)

actual = float(g)
rel_err = abs(actual - expected) / abs(expected)

print(f"x: {x_value!r}")
print(f"forward value: {float(y)!r}")
print(f"tensorflow gradient: {actual!r}")
print(f"expected gradient: {expected!r}")
print(f"relative error: {rel_err:.3e}")

if rel_err > rtol:
    print("BUG REPRODUCED: tf.nn.tanh loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
x: -19.061547465398498
forward value: -1.0
tensorflow gradient: 0.0
expected gradient: 1.1102230246251531e-16
relative error: 1.000e+00
BUG REPRODUCED: tf.nn.tanh loses a finite first-derivative signal
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun, Thank you for reporting this issue with `tf.nn.tanh`'s first-derivative calculation for `float64` input. The provided reproduction steps and detailed explanation clearly demonstrate the discrepancy between the expected and actual gradient values. We appreciate you including the standalone code, which makes it easy to verify the behavior. We will investigate this further.

**Comment by harshaljanjani** [comment]
Looking into this, thanks for the report! I'll get back with my findings.

**Comment by harshaljanjani** [comment]
Hi there, thanks for the bug report. I checked across repos for what choices were made and benchmarked some stuff, and this is what I could find:

→ <ins>PyTorch</ins>: Still output-driven, [tanh_backward(grad, result)](https://github.com/pytorch/pytorch/blob/main/tools/autograd/derivatives.yaml#L1788) is handed the output, so it should return 0 here too
→ <ins>JAX</ins>: [A commit change considered this a numerical bug](https://github.com/jax-ml/jax/commit/893b483e3f130bad7db8a0858dc229033e9ff9e1) and switched to `4 * sigmoid(2x) * sigmoid(-2x)`, but [only under accuracy=HIGHEST](https://github.com/jax-ml/jax/blob/main/jax/_src/lax/lax.py#L4612) since going back to `x` gives up the forward reuse. The default still gives 0.0, while `HIGHEST` gives 1.1102230246251531e-16
→ <ins>TensorFlow main as of today</ins>: [_TanhGrad](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L851) passes `op.outputs[0]` into the [1 - output * output](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cwise_ops_gradients.h#L31) kernel, so once `tanh(x)` rounds to `-1.0`, it can only return `0` or `2.220446049250313e-16`
→ <ins>Cost:</ins> The formula that JAX uses, `4 * sigmoid(2x) * sigmoid(-2x)`, is 1.5x to 2.1x slower (CPU) than the fused kernel across 1e5 to 1e7 elements in both dtypes, and about 1.6x on a full tanh forward + backward, which I'm assuming is why JAX gated it

My hunch is that TF does have precedent of shipping stable-but-slower behaviour behind an explicit opt-in, just like JAX did with the commit ([example, enable_op_determinism](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/config.py#L949), which notes it "comes at the expense of lower performance"), and XLA already carries [ResultAccuracy](https://github.com/tensorflow/tensorflow/blob/master/third_party/xla/xla/hlo/ir/hlo_instruction.h#L2051) which is analogous to what JAX's `accuracy=` lowers into, so we could follow through with this method here.

Happy to see it through to a PR either way
cc: @Kayyuri would love an internal opinion on this before I raise the PR :)

**Comment by Kayyuri** [comment]
Hi @ALinrunrun,
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/126747). Once it is merged, this issue will be resolved.

Hi @AshiteshSingh,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.
